test/src/jupyterhub_config/custom_spawner_test.py

```diff
@@ -20,6 +20,57 @@ def spawner():
     return spawner
 
 
+@patch.dict(os.environ, {
+    'JUPYTERHUB_USER_HOME': '/jupyterhub/users_home',
+    'JUPYTERHUB_CONFIG_DIR': '/etc/jupyterhub'
+})
+@patch.object(VirtualEnvSpawner, '_ensure_system_user')
+@patch.object(VirtualEnvSpawner, '_ensure_user_directory')
+@patch.object(VirtualEnvSpawner, '_ensure_user_jupyter_directory')
+@patch.object(VirtualEnvSpawner, '_ensure_virtual_environment')
+@patch.object(VirtualEnvSpawner, '_configure_environment')
+@patch.object(VirtualEnvSpawner, '_configure_notebook_dir')
+def test_start(mock_configure_notebook_dir, mock_configure_environment, mock_ensure_virtual_environment,
+               mock_ensure_user_jupyter_directory, mock_ensure_user_directory, mock_ensure_system_user,
+               spawner):
+
+    # set spawner.environment (_configure_environment is mocked, so `self.environment` won't be set by the method)
+    spawner.environment = {
+        'JUPYTERHUB_USER_HOME': '/jupyterhub/users_home',
+        'JUPYTERHUB_CONFIG_DIR': '/etc/jupyterhub'
+    }
+
+    spawner.start()
```

Comment:
I tested individual functions. I don't see any benefit of testing every function with one function call.

Reply:
The drawbacks are as above - you're binding the tests to internal implementation details, which is an anti-pattern, and makes refactoring much harder.

The benefit of testing the code via the API only vs. internals is that the tests ensure the behavior of the entire method call tree works together, whereas mocks can have bugs that make it *look* like they work together when they don't.

Reply:
it's making refactoring easier. How will it make it harder?

Reply:
If you change arguments or the method name of one of the sub methods you have to redo all the tests that touch that sub method. Those are implementation details, not behavior, and so there's no reason for the tests to change at all, since tests are supposed to test behavior and not implementation details.